packages/flutter/lib/src/services/platform_messages.dart

```diff
@@ -237,3 +250,647 @@ class PlatformMessages {
+  /// received from the platform plugin;
+  /// * an error event containing a [FormatException] for each event received
+  /// where decoding fails;
+  /// * an error event containing a [PlatformException] or [FormatException]
+  /// whenever stream setup fails (stream setup is done only when listener
+  /// count changes from 0 to 1).
+  ///
+  /// Notes for platform plugin implementers:
+  ///
+  /// Plugins must expose methods named `listen` and `cancel` suitable for
+  /// invocations by [invokeMethod]. Both methods are invoked with the specified
+  /// [arguments].
+  ///
+  /// Following the semantics of broadcast streams, `listen` will be called as
+  /// the first listener registers with the returned stream, and `cancel` when
+  /// the last listener cancels its registration. This pattern may repeat
+  /// indefinitely. Platform plugins should consume no stream-related resources
+  /// while listener count is zero.
+  Stream<dynamic> receiveBroadcastStream([dynamic arguments]) {
+    StreamController<dynamic> controller;
+    controller = new StreamController<dynamic>.broadcast(
+      onListen: () async {
+        PlatformMessages.setBinaryMessageHandler(
+          name,
+          (ByteData reply) async {
+              if (reply == null) {
+                controller.close();
+              } else {
+                try {
+                  controller.add(codec.decodeEnvelope(reply));
+                } catch (e) {
+                  controller.addError(e);
+                }
+              }
+          }
+        );
+        try {
+          await invokeMethod('listen', arguments);
+        } catch (e) {
+          PlatformMessages.setBinaryMessageHandler(name, null);
+          controller.addError(e);
+        }
+      }, onCancel: () async {
+        PlatformMessages.setBinaryMessageHandler(name, null);
+        try {
+          await invokeMethod('cancel', arguments);
+        } catch (exception, stack) {
+          FlutterError.reportError(new FlutterErrorDetails(
+            exception: exception,
+            stack: stack,
+            library: 'services library',
+            context: 'while de-activating platform stream on channel $name',
+          ));
+        }
+      }
+    );
+    return controller.stream;
+  }
+}
+
+/// A message encoding/decoding mechanism with support for method calls and
+/// enveloped replies.
+///
+/// Reply envelopes are binary messages with enough structure that the codec can
+/// distinguish between a successful reply and an error. In the former case,
+/// the codec must be able to extract the result, possibly `null`. In
+/// the latter case, the codec must be able to extract an error code string,
+/// a (human-readable) error message string, and a value providing any
+/// additional error details, possibly `null`. These data items are used to
+/// populate a [PlatformException].
+///
+/// All operations throw [FormatException], if conversion fails.
+abstract class MessageCodec {
```

Comment:
I'd break this file up into three separate files:

1. One file that has the abstract classes for MessageCodec and MethodCodec.
2. Another file that defines concrete implementations of these codecs for "standard" and "JSON".  This separation will prevent these codecs from being special children because (2) will be able to import (1) but (1) should not import (2).
3. A third file that contains the PlatformChannel class that has the pretty front-ends.